### ISSUE TITLE
Non-deterministic segmentation faults with std.mem.zeroes

### ISSUE THREAD

**Issue by fennewald** [ziglang/zig#10709]
### Zig Version

0.10.0-dev.431+2f41bd3be (x86_64)

### Steps to Reproduce

Given the following code:
```
const std = @import("std");
const dyn_size = 0x7fedfa;

fn foo() [dyn_size]u8 {
    var arr: [dyn_size]u8 = std.mem.zeroes([dyn_size]u8);
    return arr;
}

pub fn main() !void {
    const arr = foo();
    std.debug.print("Dynamic: {}\n", .{arr.len});
}
```

`zig build-exe main.zig` produces a binary that will randomly segfault.

### Expected Behavior

Not segfault, or at least segfault consistently.

### Actual Behavior

Segfaults randomly. From my testing:

- Increasing dyn_size leads to more certain segfaults.
- Decreasing dyn_size leads to less segfaults
- The given value of dyn_size was tuned to be as close to 50/50 segfaults on my machine
- The distribution of the above behavior is  exceedingly weird
- if `foo` is modified to be just `return std.mem.zeroes([dyn_size]u8)`, it seems to work without failure? -- My guess is this just the optimizer removing the allocation altogether.

During testing, my system was not under significant load, there was more than 32 idling cores and more than 58 GB of free memory, so it is unlikely this is a resource race. Also, I can `malloc` blocks orders of magnitude larger in c with no issues.

**Comment by fennewald** [comment]
From more testing, the segfault occurs within the call to `std.mem.zeros`

**Comment by leecannon** [comment]
The size of that array is ~8mib which is the default size of the stack on linux, due to that I'm sure this is a stack overflow.

I'm confused about why it is intermittent through.

**Comment by fennewald** [comment]
The output of
```
const std = @import("std");
const sys = @cImport({
    @cInclude("sys/resource.h");
});

pub fn main() !void {
    var l: sys.rlimit = .{ .rlim_cur = 0, .rlim_max = 0};
    _ = sys.getrlimit(sys.RLIMIT_STACK, &l);
    std.debug.print("Stack size:\n", .{});
    std.debug.print("\tCurrent: {}\n", .{l.rlim_cur});
    std.debug.print("\tMaximum: {}\n", .{l.rlim_max});
}
```
is a consistent:
```
Stack size:
	Current: 16777216
	Maximum: 16777216
```

However, very confusingly,
```
#include <sys/resource.h>
#include <stdio.h>

int main() {
  struct rlimit l;
  getrlimit(RLIMIT_STACK, &l);
  printf("Stack size:\n");
  printf("\tCurrent: %lu\n", l.rlim_cur);
  printf("\tMaximum: %lu\n", l.rlim_max);
  return 0;
}
```
Produces a consistent:
```
Stack size:
	Current: 8388608
	Maximum: 18446744073709551615
```

I am honestly now even more confused. Any idea what this means?

**Comment by mrakh** [comment]
> Produces a consistent:
> 
> ```
> Stack size:
> 	Current: 8388608
> 	Maximum: 18446744073709551615
> ```
> 
> I am honestly now even more confused. Any idea what this means?

18446744073709551615 is the maximum 64-bit unsigned value. It's basically just a sentinel for "infinite".

The stack is typically set up by the dynamic linker before the executable runs. So if your Zig and C executables are using different libcs (e.g. one is glibc and one is musl), or if one of them is not using libc at all (e.g. statically linked), then that could explain the discrepancy.

What's printed when you run these commands?

```sh
file ./your_zig_program
file ./your_c_program
```

**Comment by andrewrk** [comment]
If you want to bring determinism into debugging this, check out this project: https://rr-project.org/

Quick tutorial:

```
$ rr record ./main
$ rr replay
```

**Comment by fennewald** [comment]
@mrakh the output of file on the two programs:
```
stack_c: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, not stripped
stack_zig: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.0.0, with debug_info, not stripped
```

tl;dr, c is is a PIE executable, the zig is not, and the zig program also has `debug_info`

**Comment by mrakh** [comment]
Hmm. Do you have ASLR enabled? That's the only thing I can think of that could be introducing indeterminism this early into the execution.

**Comment by fennewald** [comment]
Noticed this was still open. Marking as resolved. It was stack overflow made a little weirder by library loading I think.
